### PR TITLE
Tenant Permissions : added the possibility to specify tenants via parameter

### PR DESCRIPTION
**Pull request by peternied**
### Description
This is a continuation of #819 with many thanks to @chrousto

The idea would be that this user :
```yaml
new-user:
  hash: "*************"
  reserved: false
  hidden: false
  opendistro_security_roles:
  - "role-tenant1"
  attributes:
    attribute1: "tenant1"
  static: false
```
Would only have access to the tenant tenant1 if the role role-tenant1 was defined like this :

```yaml
role-tenant1:
 reserved: false
 hidden: false
 cluster_permissions:
 - "read"
 - "cluster:monitor/nodes/stats"
 - "cluster:monitor/task/get"
 tenant_permissions:
 - tenant_patterns:
   - ${attr.internal.attribute1}
   allowed_actions:
   - "kibana_all_write"
 static: false
_meta:
 type: "roles"
 config_version: 2
```

### Issues Resolved
* #817 

### Testing
Integration tests have been created for this feature

### Check List
- [X] New functionality includes testing
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by DarshitChanpura — APPROVED** [review]

File: src/test/java/org/opensearch/security/multitenancy/test/MultitenancyTests.java
Comment (cliu123) [inline src/test/java/org/opensearch/security/multitenancy/test/MultitenancyTests.java:403]: Why `5.6.0`?

Comment (peternied) [reply src/test/java/org/opensearch/security/multitenancy/test/MultitenancyTests.java:403]: This was from how the test were originally authored, would you like a different value in its place?

**Comment by peternied** [comment]
@cliu123 Did you have any other comments around this pull request that need to be addressed?

**Review by cliu123 — APPROVED** [review]

File: src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java
Comment (cliu123) [inline src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:383]: This should be use while loop instead of `do...whilte`. Otherwise, this line could throw exception is `jsonPathScanner` doesn't have next.

**Review by cliu123 — CHANGES_REQUESTED** [review]
Most of the changes look great to me! I missed the `do...while` in last review though. This can be merged when it gets fixed.

File: src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java
Comment (peternied) [reply src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:383]: The only case this would happen is if the string was an empty string, then per the javadoc it would throw `@throws NoSuchElementException if no more tokens are available`.  This seems an appropriate failure that would only happen during test authorship, not in practice after a test has been merged.

Comment (cliu123) [reply src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:383]: Great! It's nice to identify the use cases where the exception could happen! Let's say the empty string case, if `NoSuchElementException` gets thrown, the method `findValueInJson` doesn't catch or throw it. And we are not supposed to let this exception happen here at all. It makes more sense to avoid the exception here by validating the condition before the operations unless it has to run the logic in `do` block before evaluating the condition.

Comment (peternied) [reply src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:383]: I think this is a matter of preference, I think the exception is good enough, if this becomes a problem over time we can enhance it.  Is this issue worth blocking the pull request?

File: src/test/java/org/opensearch/security/multitenancy/test/MultitenancyTests.java
Comment (cliu123) [reply src/test/java/org/opensearch/security/multitenancy/test/MultitenancyTests.java:403]: I think it looks fine here.

File: src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java
Comment (cliu123) [reply src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:383]: There's a valid case here as discussed above. When `jsonDotPath` is an empty string, `final String pathEntry = jsonPathScanner.next();` throws `NoSuchElementException` that is not even caught or thrown. I actually don't see a value of using `do...while` here. In fact, it's error-prone. 
This contribution is valuable, it's only a minor fix away from being merged.

Comment (cliu123) [reply src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:386]: Did this get marked as resolved in accident? unresolving this conversation

Comment (peternied) [reply src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:386]: This was resolved, see the new lines added above to cover this scenario

**Review by cliu123 — APPROVED** [review]

**Comment by peternied** [comment]
@DarshitChanpura @cliu123 could I get another review?

**Review by cliu123 — APPROVED** [review]

**Review by DarshitChanpura — APPROVED** [review]

**Pull request merged**
